From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Boris Krasnovskiy <boris.krasnovskiy@ezurio.com>
Date: Thu, 17 Sep 2026 00:00:00 +0000
Subject: [PATCH] readline: fix redisplay avoid pushing reprint character

In emacs keymap mode, C_REPRINT (^R / 0x12) is mapped to
em_inc_search_prev (reverse incremental history search).

Previously, rl_redisplay() pushed C_REPRINT into the input queue via
el_push(). When an application updated the prompt or called rl_redisplay()
on external events, this inadvertently injected ^R into the input stream,
triggering reverse history search ("bck:") and corrupting the interactive
line display.

Fix this by removing the el_push() of C_REPRINT and calling
rl_forced_update_display() directly.

Signed-off-by: Boris Krasnovskiy <boris.krasnovskiy@ezurio.com>
Upstream-Status: Pending
---
Index: libedit-20260512-3.1/src/readline.c
===================================================================
--- libedit-20260512-3.1.orig/src/readline.c	2026-09-28 15:42:28.728295765 -0400
+++ libedit-20260512-3.1/src/readline.c	2026-09-28 15:42:51.608565407 -0400
@@ -2179,11 +2179,12 @@
 void
 rl_redisplay(void)
 {
-	char a[2];
-	a[0] = (char)e->el_tty.t_c[TS_IO][C_REPRINT];
-	a[1] = '\0';
-	el_push(e, a);
-	rl_forced_update_display();
+	if (h == NULL || e == NULL)
+		rl_initialize();
+	re_clear_lines(e);
+	re_clear_display(e);
+	re_refresh(e);
+	terminal__flush(e);
 }
 
 int
